support/src/main/java/com/yubico/yubikit/support/DeviceUtil.java

```diff
@@ -0,0 +1,528 @@
+            Logger.d("Couldn't select OTP application, serial unknown");
+        }
+
+        if (version == null) {
+            Logger.d("Firmware version unknown, using 3.0.0 as a baseline");
+            version = new Version(3, 0, 0);
+        }
+
+        Logger.d("Scan for available ccid applications");
+        SmartCardProtocol protocol = new SmartCardProtocol(connection);
+        for (final CcidApplet applet : CcidApplet.values()) {
+            try {
+                protocol.select(applet.aid);
+                capabilities |= applet.capability.bit;
+            } catch (ApplicationNotAvailableException applicationNotAvailableException) {
+                Logger.d("Missing applet " + applet.name() + ", capability " + applet.capability.name());
+            } catch (IOException ioException) {
+                Logger.e("IOException selecting applet " + applet.name() + ", capability " + applet.capability.name(), ioException);
+            }
+        }
+
+        if (((interfaces & UsbInterface.FIDO) != 0) || version.isAtLeast(3, 3, 0)) {
+            capabilities |= U2F.bit;
+        }
+
+        Map<Transport, Integer> supportedCapabilities = new EnumMap<>(Transport.class);
+        supportedCapabilities.put(Transport.USB, capabilities);
+        supportedCapabilities.put(Transport.NFC, capabilities);
+
+        return new DeviceInfo(
+                new DeviceConfig.Builder().build(),
+                serial,
+                version,
+                FormFactor.UNKNOWN,
+                supportedCapabilities,
+                false,
+                false,
+                false);
+
+    }
+
+    static DeviceInfo readInfoOtp(OtpConnection connection, YubiKeyType keyType, int interfaces)
+            throws IOException {
+
+        ManagementSession managementSession = null;
+        YubiOtpSession otpSession = null;
+
+        int serial = 0;
+        Version version;
+
+        try {
+            managementSession = new ManagementSession(connection);
+        } catch (ApplicationNotAvailableException ignored) {
+            // we could not get the management session for this connection
+            // we try to get the YubiOtpSession
+            otpSession = new YubiOtpSession(connection);
+        }
+
+        // Retry during potential reclaim timeout period (~3s).
+        for (int i = 0; i < 8; i++) {
+            try {
+                if (otpSession == null) {
+                    try {
+                        return managementSession.getDeviceInfo();
+                    } catch (CommandException ignored) {
+                        // can be caused by retained state
+                        otpSession = new YubiOtpSession(connection);
+                    }
+                    serial = otpSession.getSerialNumber();
+                }
+            } catch (CommandException commandException) {
+                // can be caused by retained state
+                try {
+                    Thread.sleep(500);
```

Comment:
There's an optimization that I added in ykman to avoid this in certain cases: https://github.com/Yubico/yubikey-manager/blob/next/yubikit/support.py#L177